upgrade/scripts/k8s/upgrade_k8s.sh

```diff
@@ -0,0 +1,328 @@
+pause["1.29"]="3.9"
+pause["1.30"]="3.9"
+pause["1.31"]="3.10"
+pause["1.32"]="3.10"
+
+function prefix() {
+  echo -n "$(date --iso-8601=seconds) $(basename "$0")"
+}
+
+function fix_sysconfig() {
+  # fix_sysconfig $host
+  # Where 'host' is a master or worker host. For example, ncn-m001
+  NODE_IP=$(ssh $1 "dig +short \$(hostname).nmn | grep -v -E '^;;'")
+
+  ssh "$1" "cat > /etc/sysconfig/kubelet <<EOF
+KUBELET_EXTRA_ARGS=\"--node-ip ${NODE_IP}\"
+EOF"
+}
+
+function update_kubelet_config_cm() {
+  # First, update kubelet-config configmap to add containerRuntimeEndpoint. This
+  # is necessary because --container-runtime-endpoint was deprecated as a kubelet
+  # command line argument in K8s 1.27.
+  workdir="$(mktemp -d)"
+  kubectl get configmap kubelet-config -n kube-system -o yaml > "${workdir}/kubelet-config.yaml"
+  yq4 eval -P '.data.kubelet' "${workdir}/kubelet-config.yaml" > "${workdir}/kubelet-config-kubelet.yaml"
+  yq4 eval -i -P '.containerRuntimeEndpoint = "unix:///run/containerd/containerd.sock"' "${workdir}/kubelet-config-kubelet.yaml"
+
+  # Merge our kubelet config back into the kubelet-config manifest.
+  if IFS= read -rd '' -a kubelet_config; then
+    :
+  fi <<< "$(cat "${workdir}/kubelet-config-kubelet.yaml")"
+  kubelet_config=$kubelet_config yq4 eval -i '.data.kubelet = strenv(kubelet_config)' "${workdir}/kubelet-config.yaml"
+
+  # Update the kubelet-config configmap.
+  kubectl -n kube-system apply -f "${workdir}/kubelet-config.yaml"
+}
+
+function migrate_kubeadmcfg_yaml() {
+  # migrate_kubeadmcfg_yaml ncn-m001
+  # Given master host, run kubeadm config migrate on /etc/kubernetes/kubeadmcfg.yaml
+  host=$1
+  yaml_file="/etc/kubernetes/kubeadmcfg.yaml"
+  workdir="$(mktemp -d)"
+  echo "$(prefix) INFO Working directory for $host is $workdir"
+  old_yaml_file="${workdir}/v1beta3-kubeadmcfg.yaml"
+  new_yaml_file="${workdir}/v1beta4-kubeadmcfg.yaml"
+  # Make a copy of yaml file
+  ssh "$host" "mkdir -p $workdir; cp ${yaml_file} ${old_yaml_file}"
+  # Migrate v1beta3 config to v1beta4 config
+  ssh $host kubeadm config migrate --old-config ${old_yaml_file} | yq4 '. | select(.kind == "ClusterConfiguration") | pick(["apiVersion", "kind", "kubernetesVersion", "caCertificateValidityPeriod", "certificateValidityPeriod", "etcd"])' > ${new_yaml_file}
+  # The new_yaml_file is located on the calling host
+  # Update cert validity period to 3 years
+  yq4 eval -i -P '.certificateValidityPeriod = "26280h0m0s"' ${new_yaml_file}
+  # Copy new_yaml_file to $host:$yaml_file
+  scp $new_yaml_file $host:${yaml_file}
+}
+
+masters=$(grep -oP 'ncn-m\d+' /etc/hosts | sort -u)
+workers=$(grep -oP 'ncn-w\d+' /etc/hosts | sort -u)
+
+# We'll only upgrade the kubelet on the masters and the workers to the target version
+# which is the last version in the KUBERNETES_VERSIONS list
+target_version="${KUBERNETES_VERSIONS[-1]}"
+v_target_version="v${target_version}"
+minor_target_version=$(echo "$target_version" | cut -d'.' -f1-2)
```

Comment:
You can strip off the last component right in the shell:
```
[HPE-CG403QV9LF] t=1.32.5
[HPE-CG403QV9LF] echo ${t%.*}
1.32
```

```suggestion
minor_target_version=${target_version%.*}
```